vdr/store/bbolt.go

```diff
@@ -0,0 +1,305 @@
+package store
+
+import (
+	"encoding/json"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/crypto/hash"
+	vdr "github.com/nuts-foundation/nuts-node/vdr/types"
+	"go.etcd.io/bbolt"
+)
+
+type documentVersion struct {
+	Document did.Document
+	Metadata vdr.DocumentMetadata
+}
+
+type documentVersionList struct {
+	Deactivated bool
+	Versions    []hash.SHA256Hash
+}
+
+func (entry documentVersionList) Latest() hash.SHA256Hash {
+	return entry.Versions[len(entry.Versions)-1]
+}
+
+var (
+	documentsBucket = []byte("vdrDocuments")
+	versionsBucket  = []byte("vdrDocumentVersions")
+)
+
+type bboltStore struct {
+	db *bbolt.DB
+}
+
+// NewBBoltStore returns an instance of a BBolt based VDR store
+func NewBBoltStore(db *bbolt.DB) vdr.Store {
+	return &bboltStore{db: db}
+}
+
+func (store *bboltStore) getDocumentKey(id did.DID, hash hash.SHA256Hash) []byte {
+	return append([]byte(id.String()), hash.Slice()...)
+}
+
+func (store *bboltStore) storeDocument(tx *bbolt.Tx, document did.Document, metadata vdr.DocumentMetadata) error {
+	documents, err := tx.CreateBucketIfNotExists(documentsBucket)
+	if err != nil {
+		return err
+	}
+
+	data, err := json.Marshal(documentVersion{
+		Document: document,
+		Metadata: metadata,
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := documents.Put(store.getDocumentKey(document.ID, metadata.Hash), data); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (store *bboltStore) getDocumentVersion(bucket *bbolt.Bucket, id did.DID, hash hash.SHA256Hash) (*documentVersion, error) {
+	data := bucket.Get(store.getDocumentKey(id, hash))
+	if data == nil {
+		return nil, nil
+	}
+
+	result := &documentVersion{}
+
+	if err := json.Unmarshal(data, result); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// Iterate loops over all the latest versions of the stored DID Documents and applies fn
+func (store *bboltStore) Iterate(fn vdr.DocIterator) error {
+	return store.db.View(func(tx *bbolt.Tx) error {
```

Comment:
make sure not to call any BBolt functions from the `fn` callback. If probable, you can reuse the transaction-from-context concept from `Network` in `network/dag/bbolt_context.go`.

Reply:
I'll check if we do that somewhere